coding/geometry_coding.hpp

```diff
@@ -28,9 +29,61 @@ using InDeltasT = array_read<uint64_t>;
 using OutPointsT = array_write<m2::PointU>;
 using OutDeltasT = array_write<uint64_t>;
 
-uint64_t EncodeDelta(m2::PointU const & actual, m2::PointU const & prediction);
+// Stores the difference of two points to a single unsigned 64-bit integer.
+// It is not recommended to use this function: consider EncodePointDelta instead.
+uint64_t EncodePointDeltaAsUint(m2::PointU const & actual, m2::PointU const & prediction);
 
-m2::PointU DecodeDelta(uint64_t delta, m2::PointU const & prediction);
+m2::PointU DecodePointDeltaFromUint(uint64_t delta, m2::PointU const & prediction);
+
+// Writes the difference of two 2d vectors to sink.
+template <typename Sink>
+void EncodePointDelta(Sink & sink, m2::PointU const & curr, m2::PointU const & next)
+{
+  auto const dx = base::asserted_cast<int32_t>(next.x) - base::asserted_cast<int32_t>(curr.x);
+  auto const dy = base::asserted_cast<int32_t>(next.y) - base::asserted_cast<int32_t>(curr.y);
+  WriteVarInt(sink, dx);
+  WriteVarInt(sink, dy);
+}
+
+// Writes the difference of two 2d vectors to sink. The vector |next|
+// must have both its coordinates greater than or equal to those
+// of |curr|. While this condition is unlikely when encoding polylines,
+// the function may be useful when encoding rectangles, in particular
+// bounding boxes of shapes.
+template <typename Sink>
+void EncodePositivePointDelta(Sink & sink, m2::PointU const & curr, m2::PointU const & next)
```

Comment:
NonNegative :)

Reply:
I though about this (see the comment to function) but won't it in fact hard readability?

Reply:
Is it possible to subtract one PointU from another? If so, why do you need this method at all? I mean, just create a method that'll write to sink components of PointU as VarUints.

Reply:
Note the paranoid checks.

Reply:
And the asserts.
While this function looks superfluous, I thought it's better to leave it for now so that someone who needs to serialize bounding boxes stumbles upon it.

Reply:
I mean, why do you need to pass two arguments? Just pass a single argument - delta. So it'll be responsibility of the caller side to check that delta is non-negative (as it happens in bounding boxes).

Reply:
In this case the method looks even more superfluous.
I've returned it back to the private sections in cities_boundaries_serdes for now.